builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,188 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
 	}
 	return config, nil
 }
+
+type ClientCertificateGenerator struct {
+	CommonNameTemplate string `mapstructure:"common_name_template,omitempty"`
+	CAPrivateKey       string `mapstructure:"ca_private_key,omitempty"`
+	CACert             string `mapstructure:"ca_cert,omitempty"`
+	KeyType            string `mapstructure:"key_type,omitempty"`
+	KeyBits            int    `mapstructure:"key_bits,omitempty"`
+	SignatureBits      int    `mapstructure:"signature_bits,omitempty"`
+
+	parsedCABundle *certutil.ParsedCertBundle
+	cnProducer     template.StringTemplate
+}
+
+// newClientCertificateGenerator returns a new ClientCertificateGenerator
+// using the given config. Default values will be set on the returned
+// ClientCertificateGenerator if not provided in the config.
+func newClientCertificateGenerator(config map[string]interface{}) (ClientCertificateGenerator, error) {
+	var cg ClientCertificateGenerator
+	if err := mapstructure.WeakDecode(config, &cg); err != nil {
+		return cg, err
+	}
+
+	switch cg.KeyType {
+	case "rsa":
+		switch cg.KeyBits {
+		case 0, 2048, 3072, 4096:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ec":
+		switch cg.KeyBits {
+		case 0, 224, 256, 384, 521:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ed25519":
+		// key_bits ignored
+	default:
+		return cg, fmt.Errorf("invalid key_type")
+	}
+
+	switch cg.SignatureBits {
+	case 0, 256, 384, 512:
+	default:
+		return cg, fmt.Errorf("invalid signature_bits")
+	}
+
+	if cg.CommonNameTemplate == "" {
+		return cg, fmt.Errorf("missing required common_name_template")
+	}
+
+	// Validate the common name template
+	t, err := template.NewTemplate(template.Template(cg.CommonNameTemplate))
+	if err != nil {
+		return cg, fmt.Errorf("failed to create template: %w", err)
+	}
+
+	_, err = t.Generate(dbplugin.UsernameMetadata{})
+	if err != nil {
+		return cg, fmt.Errorf("invalid common_name_template: %w", err)
+	}
+	cg.cnProducer = t
+
+	if cg.CACert == "" {
+		return cg, fmt.Errorf("missing required ca_cert")
+	}
+	if cg.CAPrivateKey == "" {
+		return cg, fmt.Errorf("missing required ca_private_key")
+	}
+	parsedBundle, err := certutil.ParsePEMBundle(strings.Join([]string{cg.CACert, cg.CAPrivateKey}, "\n"))
+	if err != nil {
+		return cg, err
+	}
+	if parsedBundle.PrivateKey == nil {
+		return cg, fmt.Errorf("private key not found in the PEM bundle")
+	}
+	if parsedBundle.PrivateKeyType == certutil.UnknownPrivateKey {
+		return cg, fmt.Errorf("unknown private key found in the PEM bundle")
+	}
+	if parsedBundle.Certificate == nil {
+		return cg, fmt.Errorf("certificate not found in the PEM bundle")
+	}
+	if !parsedBundle.Certificate.IsCA {
```

Comment:
Strictly I think you should also validate that `BasicConstraintsValid` is asserted, too: https://pkg.go.dev/crypto/x509#Certificate

Reply:
Thanks! Done in https://github.com/hashicorp/vault/pull/20425/commits/02ef08ed3f8f74e54b60d0020caebfbb63281feb.

Reply:
Ah, I suppose this is equivalent. :laughing: I'm used to reading:

```
	if !parsedBundle.Certificate.BasicConstraintsValid || !parsedBundle.Certificate.IsCA {
	...
	}
```

but this works :-)